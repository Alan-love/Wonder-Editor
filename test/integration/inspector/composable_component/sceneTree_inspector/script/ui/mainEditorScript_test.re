open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("MainEditorScript", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());
      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test set currentSceneTreeNode", () => {
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstCubeToBeCurrentSceneTreeNode,
        );

        MainEditorInspectorAddComponentTool.addScriptComponent();
      });

      describe("test add script event function", () => {
        test("if has no script event function asset, warn", () => {
          let warn =
            createMethodStubWithJsObjSandbox(
              sandbox,
              ConsoleTool.console,
              "warn",
            );

          MainEditorScriptTool.addScriptEventFunction(
            ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
            ~send=SinonTool.createOneLengthStub(sandbox^),
            (),
          );

          warn |> expect |> toCalledOnce;
        });
        test("if has no unused script event function asset, warn", () => {
          let warn =
            createMethodStubWithJsObjSandbox(
              sandbox,
              ConsoleTool.console,
              "warn",
            );
          let assetTreeData =
            MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
          let addedNodeId = MainEditorAssetIdTool.getNewAssetId();
          MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

          MainEditorScriptTool.addScriptEventFunction(
            ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
            ~send=SinonTool.createOneLengthStub(sandbox^),
            (),
          );
          MainEditorScriptTool.addScriptEventFunction(
            ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
            ~send=SinonTool.createOneLengthStub(sandbox^),
            (),
          );

          warn |> expect |> toCalledOnce;
        });

        describe("show script event function group for add", () =>
          test("show all unused ones", () => {
            let assetTreeData =
              MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

            let state =
              MainEditorScriptTool.getUpdateState(
                ~state=
                  MainEditorScriptTool.buildState(
                    ~currentScript=
                      GameObjectTool.getCurrentSceneTreeNodeScript(),
                    (),
                  ),
                ~func=
                  MainEditorScriptTool.addScriptEventFunction(
                    ~script=GameObjectTool.getCurrentSceneTreeNodeScript(),
                    (),
                  ),
              );

            BuildComponentTool.renderScriptComponent(~state, ~sandbox, ())
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );

        describe("test remove script event function", () => {
          let _prepare = () => {
            let script = GameObjectTool.getCurrentSceneTreeNodeScript();
            let assetTreeData =
              MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
            let addedNodeId = MainEditorAssetIdTool.getNewAssetId();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

            MainEditorScriptTool.addScriptEventFunction(
              ~script,
              ~send=SinonTool.createOneLengthStub(sandbox^),
              (),
            );

            let eventFunctionName =
              ScriptEventFunctionInspectorTool.getEventFunctionName(
                addedNodeId,
              )
              |> StateLogicService.getEditorState;

            (script, eventFunctionName);
          };

          test("remove from script", () => {
            let (script, eventFunctionName) = _prepare();

            MainEditorScriptTool.removeScriptEventFunction(
              ~script,
              ~eventFunctionName,
              (),
            );

            ScriptEngineService.hasScriptEventFunctionData(
              script,
              eventFunctionName,
            )
            |> StateLogicService.getEngineStateToGetData
            |> expect == false;
          });
          test("dispatch inspector", () => {
            let (script, eventFunctionName) = _prepare();
            let dispatchFunc = SinonTool.createOneLengthStub(sandbox^);

            MainEditorScriptTool.removeScriptEventFunction(
              ~script,
              ~eventFunctionName,
              ~dispatchFunc,
              (),
            );

            let dispatchedAction =
              dispatchFunc |> getCall(0) |> getArgs |> List.hd;
            ReactTool.getDispatchUpdateActionArr(dispatchedAction)
            |> expect == [|UpdateStore.Inspector|];
          });
        });
      });

      describe("test change script event function", () => {
        describe("show script event function group for change", () =>
          test("should contain self and all unused ones", () => {
            let script = GameObjectTool.getCurrentSceneTreeNodeScript();
            let assetTreeData =
              MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

            MainEditorScriptTool.addScriptEventFunction(
              ~script,
              ~send=SinonTool.createOneLengthStub(sandbox^),
              (),
            );

            let state =
              MainEditorScriptTool.getUpdateState(
                ~state=
                  MainEditorScriptTool.buildState(~currentScript=script, ()),
                ~func=
                  MainEditorScriptTool.sendShowScriptEventFunctionGroupForChange(
                    ~scriptEventFunctionNodeId=
                      MainEditorScriptTool.getScriptAllEventFunctionNodeIds(
                        script,
                      )
                      |> StateLogicService.getStateToGetData
                      |> ArrayService.unsafeGetFirst,
                    ~script,
                    (),
                  ),
              );

            BuildComponentTool.renderScriptComponent(~state, ~sandbox, ())
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );

        describe("test change", () =>
          test("script event function group should be sorted", () => {
            let script = GameObjectTool.getCurrentSceneTreeNodeScript();
            let assetTreeData =
              MainEditorAssetTreeTool.BuildAssetTree.buildEmptyAssetTree();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();
            MainEditorAssetHeaderOperateNodeTool.addScriptEventFunction();

            MainEditorScriptTool.addScriptEventFunction(
              ~script,
              ~send=SinonTool.createOneLengthStub(sandbox^),
              (),
            );

            let targetScriptEventFunctionNodeId =
              MainEditorScriptTool.getUnUsedScriptEventFunctionNodeIds(script)
              |> StateLogicService.getStateToGetData
              |> ArrayService.unsafeGetFirst;

            let state =
              MainEditorScriptTool.getUpdateState(
                ~state=
                  MainEditorScriptTool.buildState(~currentScript=script, ()),
                ~func=
                  MainEditorScriptTool.handleChangeScriptEventFunction(
                    ~script,
                    ~currentScriptEventFunctionNodeId=
                      MainEditorScriptTool.getScriptAllEventFunctionNodeIds(
                        script,
                      )
                      |> StateLogicService.getStateToGetData
                      |> ArrayService.unsafeGetFirst,
                    ~targetScriptEventFunctionNodeId,
                  ),
              );

            let state =
              MainEditorScriptTool.getUpdateState(
                ~state=
                  MainEditorScriptTool.buildState(~currentScript=script, ()),
                ~func=
                  MainEditorScriptTool.sendShowScriptEventFunctionGroupForChange(
                    ~scriptEventFunctionNodeId=targetScriptEventFunctionNodeId,
                    ~script,
                    (),
                  ),
              );

            BuildComponentTool.renderScriptComponent(~state, ~sandbox, ())
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
      });
    });
  });