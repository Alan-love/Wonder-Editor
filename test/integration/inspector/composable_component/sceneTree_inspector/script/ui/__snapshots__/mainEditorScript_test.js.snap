// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorScript test set currentSceneTreeNode test script attribute test add script attribute show script attribute group for add show all unused ones 1`] = `
<article
  className="wonder-inspector-scriptAttribute"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Attribute Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-attribute"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Attribute
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script attribute asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script Attribute
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="scriptAttribute-fields"
        />
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script attribute
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Add Script Attribute
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script Attribute
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script Attribute 1
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script attribute test change script attribute show script attribute group for change should contain self and all unused ones 1`] = `
<article
  className="wonder-inspector-scriptAttribute"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Attribute Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-attribute"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Attribute
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script attribute asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script Attribute
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="scriptAttribute-fields"
        />
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script attribute
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Change Script Attribute
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script Attribute
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script Attribute 1
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script attribute test change script attribute test change script attribute group should be sorted 1`] = `
<article
  className="wonder-inspector-scriptAttribute"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Attribute Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-attribute"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Attribute
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script attribute asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script Attribute 1
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="scriptAttribute-fields"
        />
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script attribute
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Change Script Attribute
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script Attribute
        </div>
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script Attribute 1
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script Attribute 2
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script attribute test script attribute field show attribute's fields 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-scriptFunction"
  >
    <div
      className="inspector-component"
    >
      <div
        className="component-title"
        title="MeshRenderer"
      >
        Script Function Group
      </div>
      <hr />
      <div
        className="component-content"
      >
        <button
          className="addable-btn"
          onClick={[Function]}
        >
          Add script event function
        </button>
      </div>
    </div>
  </article>
  <article
    className="wonder-inspector-scriptAttribute"
  >
    <div
      className="inspector-component"
    >
      <div
        className="component-title"
        title="MeshRenderer"
      >
        Script Attribute Group
      </div>
      <hr />
      <div
        className="component-content"
      >
        <div
          className="wonder-script-attribute"
        >
          <div
            className="component-header"
          >
            <div
              className="header-title"
            >
              Script Attribute
            </div>
            <div
              className="header-close"
            >
              <img
                onClick={[Function]}
                src="./public/img/close.png"
              />
            </div>
          </div>
          <div
            className="inspector-item"
          >
            <div
              className="item-header"
              title="script attribute asset in use"
            >
              Name
            </div>
            <div
              className="item-content"
            >
              <div
                className="inspector-select"
              >
                <div
                  className="select-name"
                  onClick={[Function]}
                >
                  New Script Attribute
                </div>
                <div
                  className="select-img"
                  onClick={[Function]}
                >
                  <img
                    src="./public/img/select.png"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="scriptAttribute-fields"
          >
            <div
              className="scriptAttribute-field"
            >
              <article
                className="inspector-item wonder-float-input"
              >
                <div
                  className="item-header component-label"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  title="New Script Attribute Field"
                >
                  New Script Attribute Field
                </div>
                <div
                  className="item-content"
                >
                  <input
                    className="input-component float-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value="0"
                  />
                </div>
              </article>
            </div>
          </div>
        </div>
        <button
          className="addable-btn"
          onClick={[Function]}
        >
          Add script attribute
        </button>
      </div>
    </div>
  </article>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script attribute test script attribute field test set field's defaultValue test show attribute's fields 1`] = `
<article
  className="wonder-inspector-script"
>
  <article
    className="wonder-inspector-scriptFunction"
  >
    <div
      className="inspector-component"
    >
      <div
        className="component-title"
        title="MeshRenderer"
      >
        Script Function Group
      </div>
      <hr />
      <div
        className="component-content"
      >
        <button
          className="addable-btn"
          onClick={[Function]}
        >
          Add script event function
        </button>
      </div>
    </div>
  </article>
  <article
    className="wonder-inspector-scriptAttribute"
  >
    <div
      className="inspector-component"
    >
      <div
        className="component-title"
        title="MeshRenderer"
      >
        Script Attribute Group
      </div>
      <hr />
      <div
        className="component-content"
      >
        <div
          className="wonder-script-attribute"
        >
          <div
            className="component-header"
          >
            <div
              className="header-title"
            >
              Script Attribute
            </div>
            <div
              className="header-close"
            >
              <img
                onClick={[Function]}
                src="./public/img/close.png"
              />
            </div>
          </div>
          <div
            className="inspector-item"
          >
            <div
              className="item-header"
              title="script attribute asset in use"
            >
              Name
            </div>
            <div
              className="item-content"
            >
              <div
                className="inspector-select"
              >
                <div
                  className="select-name"
                  onClick={[Function]}
                >
                  New Script Attribute
                </div>
                <div
                  className="select-img"
                  onClick={[Function]}
                >
                  <img
                    src="./public/img/select.png"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="scriptAttribute-fields"
          >
            <div
              className="scriptAttribute-field"
            >
              <article
                className="inspector-item wonder-float-input"
              >
                <div
                  className="item-header component-label"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  title="New Script Attribute Field"
                >
                  New Script Attribute Field
                </div>
                <div
                  className="item-content"
                >
                  <input
                    className="input-component float-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value="1.1"
                  />
                </div>
              </article>
            </div>
          </div>
        </div>
        <button
          className="addable-btn"
          onClick={[Function]}
        >
          Add script attribute
        </button>
      </div>
    </div>
  </article>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script event function test add script event function show script event function group for add show all unused ones 1`] = `
<article
  className="wonder-inspector-scriptFunction"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Function Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-function"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Function
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script event function asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script EventFunction
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script event function
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Add Script Event Function
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script EventFunction
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script EventFunction 1
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script event function test change script event function show script event function group for change should contain self and all unused ones 1`] = `
<article
  className="wonder-inspector-scriptFunction"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Function Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-function"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Function
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script event function asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script EventFunction
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script event function
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Change Script Event Function
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script EventFunction
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script EventFunction 1
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`MainEditorScript test set currentSceneTreeNode test script event function test change script event function test change script event function group should be sorted 1`] = `
<article
  className="wonder-inspector-scriptFunction"
>
  <div
    className="inspector-component"
  >
    <div
      className="component-title"
      title="MeshRenderer"
    >
      Script Function Group
    </div>
    <hr />
    <div
      className="component-content"
    >
      <div
        className="wonder-script-function"
      >
        <div
          className="component-header"
        >
          <div
            className="header-title"
          >
            Script Function
          </div>
          <div
            className="header-close"
          >
            <img
              onClick={[Function]}
              src="./public/img/close.png"
            />
          </div>
        </div>
        <div
          className="inspector-item"
        >
          <div
            className="item-header"
            title="script event function asset in use"
          >
            Name
          </div>
          <div
            className="item-content"
          >
            <div
              className="inspector-select"
            >
              <div
                className="select-name"
                onClick={[Function]}
              >
                New Script EventFunction 1
              </div>
              <div
                className="select-img"
                onClick={[Function]}
              >
                <img
                  src="./public/img/select.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        className="addable-btn"
        onClick={[Function]}
      >
        Add script event function
      </button>
    </div>
  </div>
  <div
    className="select-component-content"
  >
    <div
      className="select-component-item"
    >
      <div
        className="select-item-header"
      >
        Change Script Event Function
      </div>
      <div
        className="select-item-body"
      >
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script EventFunction
        </div>
        <div
          className="select-item-content select-item-active"
          onClick={[Function]}
        >
          New Script EventFunction 1
        </div>
        <div
          className="select-item-content"
          onClick={[Function]}
        >
          New Script EventFunction 2
        </div>
      </div>
    </div>
    <div
      className="select-component-bg"
      onClick={[Function]}
    />
  </div>
</article>
`;
